Mathlib/LinearAlgebra/Matrix/DFT.lean

```diff
@@ -0,0 +1,274 @@
+
+/-- The IDFT matrix is the scaled conjugate transpose of the DFT matrix-/
+theorem Wₙ_conjTranspose_eq_iWₙ :  (Wₙ n)⁻¹ = ((1:ℂ)/n) • (Wₙ n)ᴴ := by
+  funext x y
+  simp only [iWₙ_apply, smul_apply, conjTranspose_apply, Wₙ_apply, smul_eq_mul]
+  rw [star_def, ← Complex.exp_conj, ← div_mul_comm, mul_one, div_left_inj', ← star_def]
+  simp only [star_div', star_natCast, star_mul', conj_I, star_def, conj_ofReal, map_ofNat, neg_mul,
+    mul_neg]
+  ring_nf
+  exact (Nat.cast_ne_zero.2 (NeZero.ne _))
+
+/-- The DFT matrix is symmetric -/
+lemma Wₙ_transpose_eq_Wₙ : (Wₙ n)ᵀ = Wₙ n := by
+  funext a b
+  simp only [transpose_apply, Wₙ_apply]
+  ring_nf
+
+def shiftk (N : ℕ) (k : Fin N) : (Fin N → Fin N) := fun n : (Fin N) => (n - k)
+
+def shiftk_equiv {N: ℕ} [hN : NeZero N] (k : Fin N) : (Fin N) ≃ (Fin N) where
+  toFun := shiftk N k
+  invFun := shiftk N (-k)
+  left_inv := by intro x;  simp only [shiftk, sub_neg_eq_add, sub_add_cancel]
+  right_inv := by intro x; simp only [shiftk, sub_neg_eq_add, add_sub_cancel]
+
+lemma cexp_shiftk_invariant (x a b : Fin n) :
+    exp (2 * ↑π * I * (↑↑a * ((x:ℂ) - (b:ℂ))) / ↑n) =
+    Complex.exp ((2*π*I)*(a*(shiftk_equiv (b) x))/n) := by
+  rw [Complex.exp_eq_exp_iff_exists_int]
+  unfold shiftk_equiv shiftk
+  simp only [sub_neg_eq_add, Equiv.coe_fn_mk]
+  by_cases h : b ≤ x
+  · use 0
+    simp only [Int.cast_zero, zero_mul, add_zero]
+    congr
+    norm_cast
+    rw [Int.subNatNat_of_le h]
+    simp only [Fin.val_fin_le, Nat.cast_inj]
+    apply ((@Fin.coe_sub_iff_le n x b).2 h).symm
+  · push_neg at h
+    use (-a)
+    rw [mul_comm ((-a:ℤ):ℂ) (2*π*I), mul_div_assoc, mul_div_assoc, mul_div_assoc, ← mul_add,
+      mul_right_inj', Int.cast_neg, Int.cast_ofNat, ← mul_div_assoc, div_eq_iff, add_mul,
+      div_mul_cancel_of_invertible, neg_mul, ← sub_eq_add_neg, ← mul_sub, mul_eq_mul_left_iff]
+    left
+    rw [Fin.coe_sub, Nat.mod_eq, if_neg]
+    norm_cast
+    rw [← Nat.add_sub_assoc, Int.subNatNat_sub, add_comm n _, Int.subNatNat_add_add]
+
+    -- Clear plethora of side goals
+    apply (le_of_lt (Nat.lt_add_left _ _ _ (Fin.is_lt _)))
+    apply (le_of_lt (Fin.is_lt _))
+    push_neg
+    intro _
+    rw [← Nat.add_sub_assoc]
+    apply Nat.sub_lt_left_of_lt_add
+    apply (le_of_lt (Nat.lt_add_left _ _ _ (Fin.is_lt _)))
+
+    apply Nat.add_lt_add_right h _
+    apply (le_of_lt (Fin.is_lt _))
+    apply (Nat.cast_ne_zero.2 (NeZero.ne _))
+    apply (mul_ne_zero (mul_ne_zero two_ne_zero _) I_ne_zero)
+    exact_mod_cast pi_ne_zero
+
+/-- A circulant matrix is diagonalized by the IDFT DFT matrix pair -/
+theorem circulant_dft  (t : Fin n → ℂ) :
+    circulant t = (Wₙ n)⁻¹ ⬝ (diagonal ( dft n t)) ⬝ (Wₙ n) := by
+  letI := invWₙ n
+  apply_fun ((Wₙ n) ⬝ ·)
+  dsimp
+  rw [Matrix.mul_assoc, Matrix.mul_inv_cancel_left_of_invertible]
+  funext a b
+  simp only [diagonal_mul, mul_apply, circulant_apply, Wₙ_apply, diagonal_apply, dft]
+  -- by_cases h : a = b
```

Comment:
please remove all random commented code unless you want us to take a look at it!